### PR TITLE
Add compute budget support 

### PR DESCRIPTION
**Pull request by 0xMMBD**
Supports instructions:
* request heap frame
* set compute unit price
* request units
* set compute unit limit

Added unit tests

Added python ragger tests for nanos, nanosp, nanox

### CODE REVIEW

File: libsol/compute_budget_instruction.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "common_byte_strings.h"
+#include "util.h"
+#include "sol/parser.h"
+#include "sol/print_config.h"
+
+extern const Pubkey compute_budget_program_id;
+
+enum ComputeBudgetInstructionKind {
+    ComputeBudgetRequestUnits = 0,//Deprecated
```
Comment (jnwng) [inline libsol/compute_budget_instruction.h]: we can remove this

File: libsol/compute_budget_instruction_test.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#include <stdio.h>
+
+void test_parse_compute_budget_instruction_kind() {
+    uint8_t message[] = {0, 1, 2, 3, 4};
+
+    enum ComputeBudgetInstructionKind instruction_kind_target;
+
+    Parser parser = {message, sizeof(message)};
+
+    // ComputeBudgetRequestUnits - Deprecated
```
Comment (jnwng) [inline libsol/compute_budget_instruction_test.c]: we can remove this

File: libsol/include/sol/transaction_summary.h [hunk 3]
```diff
@@ -69,6 +72,9 @@ SummaryItem* transaction_summary_nonce_account_item();
 SummaryItem* transaction_summary_nonce_authority_item();
 SummaryItem* transaction_summary_general_item();
 
+
+
+
```
Comment (jnwng) [inline libsol/include/sol/transaction_summary.h]: remove whitespace change

File: libsol/include/sol/transaction_summary.h [hunk 4]
```diff
@@ -45,6 +47,7 @@ enum SummaryItemKind {
     SummaryItemString,
     SummaryItemTimestamp,
 };
+
```
Comment (jnwng) [inline libsol/include/sol/transaction_summary.h]: whitespace change

File: libsol/message.c [hunk 5]
```diff
@@ -108,7 +115,7 @@ int process_message_body(const uint8_t* message_body,
     // Ensure we've consumed the entire message body
     BAIL_IF(!parser_is_empty(&parser));
 
-    // If we don't know about all of the instructions, bail
+    // If we don't know about all the instructions, bail
```
Comment (jnwng) [inline libsol/message.c]: unnecessary change

File: libsol/transaction_printers.c [hunk 6]
```diff
@@ -552,6 +554,8 @@ static int print_transaction_nonce_processed(const PrintConfig* print_config,
                         print_config);
                 case ProgramIdSerumAssertOwner:
                 case ProgramIdSplMemo:
+                // If pointers are advanced before, this case won't be needed
```
Comment (jnwng) [inline libsol/transaction_printers.c]: what does this comment mean?

File: libsol/transaction_printers.c [hunk 7]
```diff
@@ -532,6 +532,8 @@ static int print_spl_associated_token_account_create_with_transfer(const PrintCo
     return 0;
 }
 
+
+
```
Comment (jnwng) [inline libsol/transaction_printers.c]: remove whitespace change

File: libsol/transaction_printers.c [hunk 8]
```diff
@@ -634,5 +638,20 @@ int print_transaction(const PrintConfig* print_config,
         infos_length--;
     }
 
+    if (infos_length > 1) {
+        // Iterate over infos and print compute budget instructions and offset pointers
+        // Handle ComputeBudget instructions first due to teh limitations of the print_transaction_nonce_processed.
```
Comment (jnwng) [inline libsol/transaction_printers.c]: not sure what this comment means

File: libsol/transaction_summary.c [hunk 9]
```diff
@@ -98,6 +98,7 @@ void transaction_summary_reset() {
     explicit_bzero(&G_transaction_summary_text, TEXT_BUFFER_LENGTH);
 }
 
+
```
Comment (jnwng) [inline libsol/transaction_summary.c]: remove whitespace changes

File: libsol/transaction_summary_test.c [hunk 10]
```diff
@@ -1,4 +1,5 @@
 #include "common_byte_strings.h"
+#include "sol/transaction_summary.h"
```
Comment (jnwng) [inline libsol/transaction_summary_test.c:2]: why the `sol/` prefix?

Comment (0xMMBD) [reply libsol/transaction_summary_test.c:2]: that header file is in different directory. I had it auto imported by my IDE as include/sol but i've changed it to the sol/... after our last discussion to make it more consistent across other files

File: libsol/transaction_printers.c (hunk 8 above)
Comment (0xMMBD) [reply libsol/transaction_printers.c]: Main problem is inside `transaction_printers.c` file. We have there hardcoded cases of instructions in one big switch statement. Because of it, handling all cases related to the ComputeBudget would introduce a lot of low quality and unreadable code. It can be solved by iterating over CB instructions first, this way we can not care about the amount nor the order.

This is one of the problems that blocked us with the token2022 extensions. To many combinations to handle

Comment (0xMMBD) [reply libsol/transaction_printers.c]: I can extend this comment if you want or just delete it, it's up to you

File: libsol/transaction_printers.c (hunk 6 above)
Comment (0xMMBD) [reply libsol/transaction_printers.c]: This is related to the comment below. `ProgramIdComputeBudget` is printed in slightly different way, so we don't want to return any print function here

File: libsol/include/sol/transaction_summary.h (hunk 3 above)
Comment (0xMMBD) [reply libsol/include/sol/transaction_summary.h]: Thanks for the comment, I think that one commit with formatting fixes was lost somewhere when i was cherry picking. I'll handle those cases

File: libsol/include/sol/transaction_summary.h [hunk 11]
```diff
@@ -20,6 +20,8 @@
 // If all _Required_ `SummaryItem`s have not been set, finalization will fail.
 
 #define NUM_GENERAL_ITEMS 11
+#define DEFAULT_COMPUTE_UNIT_LIMIT 200000
+#define COMPUTE_UNIT_PRICE_DIVIDER 1000000
```
Comment (sponomarev) [inline libsol/include/sol/transaction_summary.h]: Are those used anywhere?

File: libsol/system_instruction.c [hunk 12]
```diff
@@ -260,6 +260,7 @@ static int print_system_transfer_info(const SystemTransferInfo* info,
     SummaryItem* item;
 
     item = transaction_summary_primary_item();
+
```
Comment (sponomarev) [inline libsol/system_instruction.c]:
```suggestion
```

File: libsol/include/sol/transaction_summary.h (hunk 11 above)
Comment (0xMMBD) [reply libsol/include/sol/transaction_summary.h]: :thinking: good catch, It must been left from initial PR (It changed few times). I've removed it

**Review by sponomarev — APPROVED** [review]
LGTM!

**Review by tiago18c — APPROVED** [review]
lgtm!
